### PR TITLE
fix(@aws-amplify/api) Side effect endpoint method

### PR DESCRIPTION
**Pull request by elorzafe**
_Issue #, if available:_
Fixes: #5858
_Description of changes:_
Ignore original `endpoint` method (keeped for backwards compatibility) that has a side effect. Add `getEndpointInfo` method which searches the api configuration and returns the new type `ApiInfo`.

Previously was failing with InvalidSignatureException on an App like this.

```jsx
// App.tsx
import React, { useEffect, useState } from "react";
import logo from "./logo.svg";
import "./App.css";
import Amplify, { API } from "aws-amplify";
import awsconfig from './aws-exports';

Amplify.configure(awsconfig);

function App() {
  const [lambdaResult, setLambdaResult] = useState({});
  const [apiGetResult, setApiGetResult] = useState({});
  const [apiPostResult, setApiPostResult] = useState({});
  
  useEffect(() => {
    API.post("api1", "/items", { body: {} }).then((data) => setApiPostResult(data)).catch(err => console.warn({err}));
    API.post("lambda", "", { body: {} }).then((data) => setLambdaResult(data)).catch(err => console.warn({err}));
    API.get("api1", "/items", {}).then((data) => setApiGetResult(data)).catch(err => console.warn({err}));
    API.post("lambda", "", { body: {} }).then((data) => setLambdaResult(data)).catch(err => console.warn({err}));
  }, []);
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <pre data-test="api-get">{JSON.stringify(apiGetResult, null, 2)}</pre>
        <pre data-test="api-post">{JSON.stringify(apiPostResult, null, 2)}</pre>
        <pre data-test="lambda">{JSON.stringify(lambdaResult, null, 2)}</pre>
      </header>
    </div>
  );
}

export default App;
```

With a config like this
```javascript
/* eslint-disable */
// WARNING: DO NOT EDIT. This file is automatically generated by AWS Amplify. It will be overwritten.

const awsmobile = {
    "aws_project_region": "us-west-2",
    "aws_cognito_identity_pool_id": "us-west-2:xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
    "aws_cognito_region": "us-west-2",
    "aws_user_pools_id": "us-west-2_xxxxxxxx",
    "aws_user_pools_web_client_id": "xxxxxxxxxxx",
    "oauth": {},
    "aws_cloud_logic_custom": [
        {
            "name": "api1",
            "endpoint": "https://xxxxxx.execute-api.us-west-2.amazonaws.com/dev",
            "region": "us-west-2"
        },
        {
            name: "lambda",
            endpoint: "https://lambda.us-west-2.amazonaws.com/2015-03-31/functions/xxxxxxxxxx-dev/invocations",
            service: "lambda",
            region: "us-west-2"
          },
    ],
};


export default awsmobile;

```

With this cypress spec
```javascript
describe('API tests', () => {
    beforeEach(() => {
        cy.visit('/');
    });

    it('calling api and lambda', () => {
        cy.get('[data-test="api-get"]').contains('get call succeed!');
        cy.get('[data-test="api-post"]').contains('post call succeed!');
        cy.get('[data-test="lambda"]').contains('Hello');
    });
});
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/api-rest/src/RestAPI.ts [hunk 1]
```diff
@@ -242,6 +281,50 @@ export class RestAPIClass {
...
+		};
+
+		if (typeof apiConfig.region === 'string') {
+			response.region = apiConfig.region;
+		} else if (typeof this._options.region === 'string') {
+			response.region = this._options.region;
+		}
+
+		if (typeof apiConfig.service === 'string') {
+			response.service = apiConfig.service || 'execute-api';
```
Comment (ericclemmons) [inline packages/api-rest/src/RestAPI.ts:315]: Is `apiConfig.service` often `""`?  I'm wondering iin what situation would `execute-api` be a default here.

File: packages/api-rest/src/RestClient.ts [hunk 2]
```diff
@@ -81,15 +81,29 @@ export class RestClient {
...
+		let service: string = 'execute-api';
+		let custom_header: () => {
+			[key: string]: string;
+		} = undefined;
+
+		if (typeof urlOrApiInfo === 'string') {
+			parsed_url = this._parseUrl(urlOrApiInfo);
+			url = urlOrApiInfo;
+		} else {
+			({ endpoint: url, custom_header, region, service } = urlOrApiInfo);
```
Comment (ericclemmons) [inline packages/api-rest/src/RestClient.ts:104]: This is the first time I've seen destructuring over `let` variables!

File: packages/api-rest/src/RestAPI.ts [hunk 3]
```diff
@@ -204,16 +237,22 @@ export class RestAPIClass {
...
-			return Promise.reject('API ' + apiName + ' does not exist');
+		try {
+			const apiConfig = this.getEndpointInfo(apiName, path);
+
+			const cancellableToken = this._api.getCancellableToken();
+
+			const initParams = Object.assign({}, init);
+			initParams.cancellableToken = cancellableToken;
+
+			const responsePromise = this._api.head(apiConfig, initParams);
```
Comment (ericclemmons) [inline packages/api-rest/src/RestAPI.ts:248]: The other methods call this `apiInfo`.  Not a bug, but may be inconsistent?

```suggestion
			const apiInfo = this.getEndpointInfo(apiName, path);

			const cancellableToken = this._api.getCancellableToken();

			const initParams = Object.assign({}, init);
			initParams.cancellableToken = cancellableToken;

			const responsePromise = this._api.head(apiInfo, initParams);
```

**Comment by elorzafe** [comment]
> This looks good @elorzafe! Nothing to block over, but I did want to confirm my understanding:
> 
> After reading @manueliglesias's comment ([#5858 (comment)](https://github.com/aws-amplify/amplify-js/issues/5858#issuecomment-632383146)), I understand the "why" better: _the methods customers interact with (e.g. `API.post('myApi', 'path', ...)`) "lose" their `region`/`service` values in the underlying `_api.*` calls.
> 
> This PR makes those calls accept the previous signature (a string) as well as the new `ApiInfo` struct so `region` and `service` make their way through the entire callstack.
> 
> Did I get that right? 😅

Exactly! internal classes are exported also so I tried to not brake those. But the category will be using the `getEndpointInfo` that return `ApiInfo`.

Thanks @ericclemmons !

File: packages/api-rest/src/types/index.ts [hunk 4]
```diff
@@ -64,3 +64,10 @@ export interface apiOptions {
 	endpoints: object;
 	credentials?: object;
 }
+
+export type ApiInfo = {
+	endpoint: string;
+	region?: string;
+	service?: string;
+	custom_header?: () => { [key: string]: string };
```
Comment (manueliglesias) [inline packages/api-rest/src/types/index.ts:72]: Shouldn't this be plural?

```suggestion
	custom_headers?: () => { [key: string]: string };
```

Comment (manueliglesias) [reply packages/api-rest/src/types/index.ts:72]: If this name bubbles up all the way to the public API, we might have to live with it in its singular form as to not break consumers.

File: packages/api-rest/src/RestAPI.ts (hunk 1 above)
Comment (elorzafe) [reply packages/api-rest/src/RestAPI.ts:315]: Good question, 'execute-api' is the one that is used by api gateway. Is the default and most common option.

File: packages/api-rest/src/RestClient.ts (hunk 2 above)
Comment (elorzafe) [reply packages/api-rest/src/RestClient.ts:104]: :)

File: packages/api-rest/src/RestAPI.ts (hunk 3 above)
Comment (elorzafe) [reply packages/api-rest/src/RestAPI.ts:248]: Thanks, I will rename that

File: packages/api-rest/src/RestClient.ts [hunk 5]
```diff
@@ -81,15 +81,29 @@ export class RestClient {
...
 	 * @param {string} method - Request HTTP method
 	 * @param {json} [init] - Request extra params
 	 * @return {Promise} - A promise that resolves to an object with response status and JSON data, if successful.
 	 */
-	async ajax(url: string, method: string, init) {
-		logger.debug(method + ' ' + url);
-
-		const parsed_url = this._parseUrl(url);
+	async ajax(urlOrApiInfo: string | ApiInfo, method: string, init) {
+		logger.debug(method + ' ' + urlOrApiInfo);
```
[the remark this answers is not in the document]
Comment (elorzafe) [reply packages/api-rest/src/RestClient.ts:90]: Thanks @manueliglesias I will fix that

File: packages/api-rest/src/types/index.ts (hunk 4 above)
Comment (elorzafe) [reply packages/api-rest/src/types/index.ts:72]: is all the way up :(

File: packages/api-rest/src/RestAPI.ts (hunk 1 above)
Comment (ericclemmons) [reply packages/api-rest/src/RestAPI.ts:315]: Put another way, what values of `apiConfig.service` will both be a `string` and default to `execute-api`?

`""` is the only one I'm aware of, but if that won't be the case, then `|| 'execute-api'` wouldn't be necessary here.

File: packages/api-rest/src/RestAPI.ts [hunk 6]
```diff
@@ -62,23 +63,25 @@ export class RestAPIClass {
 			});
 		}
 
-		if (!Array.isArray(opt.endpoints)) {
+		if (Array.isArray(opt.endpoints)) {
```
Comment (sammartinez) [inline packages/api-rest/src/RestAPI.ts:66]: Do we need the `!` here?

Comment (sammartinez) [reply packages/api-rest/src/RestAPI.ts:66]: As per our offline this is fine

**Pull request merged**